### ISSUE TITLE
Error while installing - Undefined reference to 'xcb_randr_get_monitors'

### ISSUE THREAD

**Issue by karma-riuk** [jaagr/polybar#1154]
Hey there :)

I am very sorry if this has already been solved but I just spent my whole afternoon trying to look up on the internet how to install polybar. I've run in a lot of 'issues' going through building and installing it, such as dependencies. I am relatively new to linux and programming so that's why it took me a while to fix.

But now, I've litteraly tried everything I found on the internet about it, but nothing works.

I think it is necessary to note that I am on Ubuntu 17.04.

## The issue

The cmake goes through very gentely, without any issues. I'll still put the output since I saw that on previous threads it seemed pretty usefull. 
After running the command `sudo cmake -DCMAKE_CXX_COMPILER=g++ ..` in the polybar/build directory [here is what comes out](https://github.com/karma-riuk/polybar-outputs/blob/master/cmake) (I just added a file to my repo in order not to spam this whole post with tones of output lines)

The problem comes up after I run the command `sudo make install` (still in the polybar/build directory)
[where the whole output is here](https://github.com/karma-riuk/polybar-outputs/blob/master/make%20install), but I'll post here is the bug report that appears so that you can see it directly:
``
[ 95%] Linking CXX executable polybar
CMakeFiles/polybar.dir/x11/extensions/randr.cpp.o: In function `xcb_randr_get_monitors_cookie_t xpp::generic::reply<xpp::randr::reply::detail::get_monitors<polybar::v3_1_0_90_g1d20df0::detail::connection_base<polybar::v3_1_0_90_g1d20df0::connection&, xpp::randr::extension, xpp::xkb::extension>, xpp::generic::checked_tag, xpp::generic::signature<xcb_randr_get_monitors_cookie_t (xcb_connection_t*, unsigned int, unsigned char), xcb_randr_get_monitors> >, polybar::v3_1_0_90_g1d20df0::detail::connection_base<polybar::v3_1_0_90_g1d20df0::connection&, xpp::randr::extension, xpp::xkb::extension>, xpp::generic::checked_tag, xpp::generic::signature<xcb_randr_get_monitors_reply_t* (xcb_connection_t*, xcb_randr_get_monitors_cookie_t, xcb_generic_error_t**), xcb_randr_get_monitors_reply>, xpp::generic::signature<xcb_randr_get_monitors_cookie_t (xcb_connection_t*, unsigned int, unsigned char), xcb_randr_get_monitors> >::cookie<polybar::v3_1_0_90_g1d20df0::detail::connection_base<polybar::v3_1_0_90_g1d20df0::connection&, xpp::randr::extension, xpp::xkb::extension>, unsigned int&, bool>(polybar::v3_1_0_90_g1d20df0::detail::connection_base<polybar::v3_1_0_90_g1d20df0::connection&, xpp::randr::extension, xpp::xkb::extension>&&, unsigned int&, bool&&)':
/home/arno/Downloads/polybar/lib/xpp/include/xpp/proto/../generic/request.hpp:99: undefined reference to `xcb_randr_get_monitors'
CMakeFiles/polybar.dir/x11/extensions/randr.cpp.o: In function `xpp::generic::iterator<polybar::v3_1_0_90_g1d20df0::detail::connection_base<polybar::v3_1_0_90_g1d20df0::connection&, xpp::randr::extension, xpp::xkb::extension>, xcb_randr_monitor_info_t, xpp::generic::signature<void (xcb_randr_monitor_info_iterator_t*), xcb_randr_monitor_info_next>, xpp::generic::signature<int (void const*), xcb_randr_monitor_info_sizeof>, xpp::generic::signature<xcb_randr_monitor_info_iterator_t (xcb_randr_get_monitors_reply_t const*), xcb_randr_get_monitors_monitors_iterator> >::operator++()':
/home/arno/Downloads/polybar/lib/xpp/include/xpp/proto/../generic/reply_iterator.hpp:169: undefined reference to `xcb_randr_monitor_info_sizeof'
/home/arno/Downloads/polybar/lib/xpp/include/xpp/proto/../generic/reply_iterator.hpp:170: undefined reference to `xcb_randr_monitor_info_next'
CMakeFiles/polybar.dir/x11/extensions/randr.cpp.o: In function `xpp::generic::reply<xpp::randr::reply::detail::get_monitors<polybar::v3_1_0_90_g1d20df0::detail::connection_base<polybar::v3_1_0_90_g1d20df0::connection&, xpp::randr::extension, xpp::xkb::extension>, xpp::generic::checked_tag, xpp::generic::signature<xcb_randr_get_monitors_cookie_t (xcb_connection_t*, unsigned int, unsigned char), xcb_randr_get_monitors> >, polybar::v3_1_0_90_g1d20df0::detail::connection_base<polybar::v3_1_0_90_g1d20df0::connection&, xpp::randr::extension, xpp::xkb::extension>, xpp::generic::checked_tag, xpp::generic::signature<xcb_randr_get_monitors_reply_t* (xcb_connection_t*, xcb_randr_get_monitors_cookie_t, xcb_generic_error_t**), xcb_randr_get_monitors_reply>, xpp::generic::signature<xcb_randr_get_monitors_cookie_t (xcb_connection_t*, unsigned int, unsigned char), xcb_randr_get_monitors> >::get(xpp::generic::checked_tag)':
/home/arno/Downloads/polybar/lib/xpp/include/xpp/proto/../generic/request.hpp:112: undefined reference to `xcb_randr_get_monitors_reply'
CMakeFiles/polybar.dir/x11/extensions/randr.cpp.o: In function `xpp::generic::iterator<polybar::v3_1_0_90_g1d20df0::detail::connection_base<polybar::v3_1_0_90_g1d20df0::connection&, xpp::randr::extension, xpp::xkb::extension>, xcb_randr_monitor_info_t, xpp::generic::signature<void (xcb_randr_monitor_info_iterator_t*), xcb_randr_monitor_info_next>, xpp::generic::signature<int (void const*), xcb_randr_monitor_info_sizeof>, xpp::generic::signature<xcb_randr_monitor_info_iterator_t (xcb_randr_get_monitors_reply_t const*), xcb_randr_get_monitors_monitors_iterator> >::iterator<polybar::v3_1_0_90_g1d20df0::detail::connection_base<polybar::v3_1_0_90_g1d20df0::connection&, xpp::randr::extension, xpp::xkb::extension>&>(polybar::v3_1_0_90_g1d20df0::detail::connection_base<polybar::v3_1_0_90_g1d20df0::connection&, xpp::randr::extension, xpp::xkb::extension>&, std::shared_ptr<xcb_randr_get_monitors_reply_t const> const&)':
/home/arno/Downloads/polybar/lib/xpp/include/xpp/proto/../generic/reply_iterator.hpp:144: undefined reference to `xcb_randr_get_monitors_monitors_iterator'
/usr/local/lib/libxcb-shm.so.0: undefined reference to `xcb_send_request_with_fds'
collect2: error: ld returned 1 exit status
bin/CMakeFiles/polybar.dir/build.make:2107: recipe for target 'bin/polybar' failed
make[2]: *** [bin/polybar] Error 1
CMakeFiles/Makefile2:816: recipe for target 'bin/CMakeFiles/polybar.dir/all' failed
make[1]: *** [bin/CMakeFiles/polybar.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
``

The output of `pkg-config --print-provides xcb-randr` is `xcb-randr = 1.13`

I repeat, I am very sorry if I am wasting your time because the answer to this problem was already given, but I just can't see to get out of all these bugs... :(

I would be very happy to give you any information you need to help me fix the install of polybar. I am really eager to make it work on my laptop because it looks beautifull so I don't want to give up!

Thank you very much in advance,
Karma

**Comment by patrick96** [comment]
Looks a lot like #952 but we weren't able to resolve that one.
What are the versions of the following packages on your system?
```
libcairo2-dev
libxcb1-dev
libxcb-ewmh-dev
libxcb-icccm4-dev
libxcb-image0-dev
libxcb-randr0-dev
libxcb-util0-dev
libxcb-xkb-dev
python-xcbgen
xcb-proto
```
You can get the version number with `dpkg -s`

**Comment by karma-riuk** [comment]
Okay, so the versions of these packages are

- libcairo2-dev `Version: 1.14.8-1`
- libxcb1-dev `Version: 1.11.1-1ubuntu1`
- libxcb-ewmh-dev `Version: 0.4.1-1ubuntu1`
- libxcb-icccm4-dev `Version: 0.4.1-1ubuntu1`
- libxcb-image0-dev `Version: 0.4.0-1build1`
- libxcb-randr0-dev `Version: 1.11.1-1ubuntu1`
- libxcb-util0-dev `Version: 0.4.0-0ubuntu3`
- libxcb-xkb-dev `Version: 1.11.1-1ubuntu1`
- python-xcbgen `Version: 1.12-1`
- xcb-proto `Version: 1.12-1`

**Comment by patrick96** [comment]
Hmm strange that `pkg-config --print-provides xcb-randr` returns `1.13` when the actual version should be `1.11`. Did you build or install xcb-randr from anywhere else?

**Comment by karma-riuk** [comment]
Well, I'm not really used to install manually dependencies from internet so maybe I've done something wrong, but I followed the instructions on [Linux From Scratch's guide](http://www.linuxfromscratch.org/blfs/view/svn/x/libxcb.html) to install the libxcb.

Maybe libxcb and libxcb-randr are two distinct things, in such case, please correct me and indicate me which version and how to install libxcb-randr properly. I would be very grateful :)

**Comment by patrick96** [comment]
Ubuntu actually provides libxcb in its own repositories, in fact you have already installed version `1.11`. Additionally you have also installed version `1.13` with the LFS instructions. The reason the build breaks this way is because you have the same package installed twice with different versions. It's probably best to remove version `1.13` of the package and reinstall `libxcb1-dev` from the repos.

**Comment by karma-riuk** [comment]
Okay, I will try this.

**Comment by patrick96** [comment]
You should be able to do `sudo make uninstall` in the directory you built `libxcb` in.

**Comment by karma-riuk** [comment]
Okay, I've done it, but now I have a new error... And [here is what comes out](https://github.com/karma-riuk/polybar-outputs/blob/master/sudo%20make%20install%20(after%20uninstalling%20libxcb%201.13))

Here are the last thew lines, just to have an idea of the bug:
`cc1plus: error: unrecognized command line option ‘-Wno-noexcept-type’ [-Werror]
cc1plus: all warnings being treated as errors
bin/CMakeFiles/polybar.dir/build.make:206: recipe for target 'bin/CMakeFiles/polybar.dir/components/bar.cpp.o' failed
make[2]: *** [bin/CMakeFiles/polybar.dir/components/bar.cpp.o] Error 1
CMakeFiles/Makefile2:816: recipe for target 'bin/CMakeFiles/polybar.dir/all' failed
make[1]: *** [bin/CMakeFiles/polybar.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2`

**Comment by patrick96** [comment]
Alright that looks great. Well not great, but now this is actually the same problem as #644 and we already have a way of solving that one. You will need to downgrade `xcb-proto` to `1.11`, we have detailed instructions for that [here](https://github.com/jaagr/polybar/wiki/Compiling#version-mismatch-between-xcb-proto-and-libxcb-randr0-dev).

Let me know, if it worked

**Comment by karma-riuk** [comment]
Unfortunately, it hasn't change anything, or at least it looks like it. [Here is the link to the 'new' error output](https://github.com/karma-riuk/polybar-outputs/blob/master/sudo%20make%20install%20(after%20changing%20the%20version%20of%20xcb-proto))

And as usual the last thew lines, just in case:
`
cc1plus: error: unrecognized command line option ‘-Wno-noexcept-type’ [-Werror]
cc1plus: all warnings being treated as errors
bin/CMakeFiles/polybar.dir/build.make:206: recipe for target 'bin/CMakeFiles/polybar.dir/components/bar.cpp.o' failed
make[2]: *** [bin/CMakeFiles/polybar.dir/components/bar.cpp.o] Error 1
CMakeFiles/Makefile2:816: recipe for target 'bin/CMakeFiles/polybar.dir/all' failed
make[1]: *** [bin/CMakeFiles/polybar.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
`

I checked the command `pkg-config --print-provides xcb-proto` before executing anything, it gave `1.13` and now, after uninstalling and installing the older one, it returns `1.11`

**Comment by patrick96** [comment]
Did you redownload the polybar repository?

**Comment by karma-riuk** [comment]
Wow! It worked!!!! 

I did as you told me, so I basically erased the polybar dir that was downloaded, I've gone through every command show in the README.md and it 'magically' worked.

Thank you very much @patrick96 for your help and for answering so quickly!

I hope it will be usefull not only for me but for people who encountered the same problem as I encountered and couldn't find enough information online.

Thanks again :)
